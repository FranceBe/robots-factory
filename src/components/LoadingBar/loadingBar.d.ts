export type LoadingBarProps = {
  timeLeft?: number
  taskTime: number
}
