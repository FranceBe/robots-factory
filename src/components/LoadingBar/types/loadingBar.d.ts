export type LoadingBarProps = {
  timeLeft?: number
  timeBase: number | null
}
