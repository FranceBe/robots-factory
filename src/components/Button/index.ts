import ButtonDefault, { Button } from 'components/Button/Button.component'

export { Button }

export default ButtonDefault
