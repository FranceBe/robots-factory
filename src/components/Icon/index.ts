import { FAIconByIconProps, Icon } from 'components/Icon/Icon.component'

export { FAIconByIconProps, Icon }
