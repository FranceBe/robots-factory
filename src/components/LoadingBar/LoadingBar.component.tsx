import { LoadingBarProps } from 'components/LoadingBar/loadingBar'
import {
  FilledBar,
  LoadingBarBackground,
  LoadingBarContainer,
  TextContainer,
} from 'components/LoadingBar/loadingBar.style'
import { getFilledBarWidth } from 'components/LoadingBar/loadingBar.utils'
import React, { useMemo, useState } from 'react'

export const LoadingBar: React.FC<LoadingBarProps> = ({ timeLeft = 0, timeBase }) => {
  const [fillBarPercent, setFillBarPercent] = useState(getFilledBarWidth(timeLeft, timeBase))

  useMemo(() => setFillBarPercent(getFilledBarWidth(timeLeft, timeBase)), [timeLeft, timeBase])

  return (
    <LoadingBarContainer data-testid={'loading-bar'}>
      <LoadingBarBackground>
        <FilledBar data-testid={'filled-bar'} filled={fillBarPercent} />
      </LoadingBarBackground>
      <TextContainer>{Math.round(timeLeft * 100) / 100} s</TextContainer>
    </LoadingBarContainer>
  )
}

export default LoadingBar
