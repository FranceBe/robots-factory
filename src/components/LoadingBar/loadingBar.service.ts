// Calculate a percentage from the timeBase and the timeLeft
// To display correctly the width of the FilledBar styled-component
export const getFilledBarWidth = (timeLeft: number, timeBase: number | null): number => {
  if (timeLeft < 0 || !timeBase) {
    return 0
  }
  const timeDifference = timeBase - timeLeft
  if (timeDifference < 0) {
    return 100
  }
  const percent = (timeDifference / timeBase) * 100
  return Math.round(percent * 100) / 100
}
