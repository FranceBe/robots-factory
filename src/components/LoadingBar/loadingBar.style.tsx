import React, { PropsWithChildren } from 'react'
import styled from 'styled-components'
import { fontSizes, palette, spaces } from 'styles/variables'

export const LoadingBarContainer = styled.div`
  height: ${spaces.medium};
  border-radius: ${spaces.x_small};
`
export const LoadingBarBackground = styled.div`
  background-color: ${palette.grey_tertiary};
  border-radius: ${spaces.x_small};
  display: flex;
  width: 100%;
  height: 100%;
`

export const TextContainer = styled.div`
  position: relative;
  top: -19px;
  left: 94%;
  font-size: ${fontSizes.small};
  z-index: 2;
  width: fit-content;
`

export const FilledBar = styled((props: PropsWithChildren<{ filled: number }>) => (
  <div {...props} />
))`
  background-color: ${palette.green_primary};
  width: ${(props) => `${props.filled}%`};
  border-bottom-left-radius: ${spaces.x_small};
  border-top-left-radius: ${spaces.x_small};
  border-bottom-right-radius: ${(props) => (props.filled === 100 ? spaces.x_small : 0)};
  border-top-right-radius: ${(props) => (props.filled === 100 ? spaces.x_small : 0)};
  height: 100%;
  transition: ${(props) => (props.filled === 0 ? 'none' : 'width 1s linear')};
`
