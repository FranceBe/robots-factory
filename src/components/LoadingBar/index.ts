import { LoadingBar } from 'components/LoadingBar/LoadingBar.component'

export { LoadingBar }
