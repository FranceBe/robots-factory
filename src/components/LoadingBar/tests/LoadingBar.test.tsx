import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { render, screen } from '@testing-library/react'
import { LoadingBar } from 'components/LoadingBar'
import { LoadingBarProps } from 'components/LoadingBar/types/loadingBar'
import React from 'react'

describe('LoadingBar component', () => {
  const initTest = (props?: Partial<LoadingBarProps>) =>
    render(<LoadingBar timeBase={100} {...props} />)

  it('should match snapshot', () => {
    const { container } = initTest()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render FilledBar with the result of getFilledBarWidth as filled property', () => {
    initTest({ timeLeft: 50 })
    // In this case, timeLeft is 50 and timeBase is 100, so the result
    // of getFilledBarWidth(50, 100) should be 50

    const filledBar = screen.getByTestId('filled-bar')
    expect(filledBar.getAttribute('filled')).toBe('50')
  })
  it('should render timeLeft as text with an "s" as second', () => {
    initTest({ timeLeft: 50 })

    expect(screen.getByText('50 s')).toBeInTheDocument()
  })
})
