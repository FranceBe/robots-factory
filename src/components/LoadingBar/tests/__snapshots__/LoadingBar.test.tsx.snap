// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingBar component should match snapshot 1`] = `
.c0 {
  height: 24px;
  border-radius: 4px;
}

.c1 {
  background-color: #D4D5D6;
  border-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
}

.c3 {
  position: relative;
  top: -19px;
  left: 94%;
  font-size: 0.875rem;
  z-index: 2;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.c2 {
  background-color: #45B69C;
  width: 100%;
  border-bottom-left-radius: 4px;
  border-top-left-radius: 4px;
  border-bottom-right-radius: 4px;
  border-top-right-radius: 4px;
  height: 100%;
}

<div
  class="c0"
  data-testid="loading-bar"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      data-testid="filled-bar"
      filled="100"
    />
  </div>
  <div
    class="c3"
  >
    0
     s
  </div>
</div>
`;
